c++/src/Timezone.cc

```diff
@@ -0,0 +1,964 @@
+                           uint64_t nameCount);
+    void parseZoneFile(const unsigned char* ptr,
+                       uint64_t sectionOffset,
+                       uint64_t fileLength,
+                       const VersionParser& version);
+    // filename
+    std::string filename;
+
+    // the version of the file
+    uint64_t version;
+
+    // the list of variants for this timezone
+    std::vector<TimezoneVariant> variants;
+
+    // the list of the times where the local rules change
+    std::vector<int64_t> transitions;
+
+    // the variant that starts at this transition.
+    std::vector<uint64_t> currentVariant;
+
+    // the variant before the first transition
+    uint64_t ancientVariant;
+
+    // the rule for future times
+    std::unique_ptr<FutureRule> futureRule;
+
+    // the last explicit transition after which we use the future rule
+    int64_t lastTransition;
+
+    // The ORC epoch time in this timezone.
+    int64_t epoch;
+  };
+
+  DIAGNOSTIC_PUSH
+  #ifdef __clang__
+    DIAGNOSTIC_IGNORE("-Wglobal-constructors")
+    DIAGNOSTIC_IGNORE("-Wexit-time-destructors")
+  #endif
+  static std::map<std::string, Timezone*> timezoneCache;
+  DIAGNOSTIC_POP
+
+  Timezone::~Timezone() {
+    // PASS
+  }
+
+  TimezoneImpl::TimezoneImpl(const std::string& _filename,
+                             const std::vector<unsigned char> buffer
+                             ): filename(_filename) {
+    parseZoneFile(&buffer[0], 0, buffer.size(), Version1Parser());
+    // Build the literal for the ORC epoch
+    // 2015 Jan 1 00:00:00
+    tm epochStruct;
+    epochStruct.tm_sec = 0;
+    epochStruct.tm_min = 0;
+    epochStruct.tm_hour = 0;
+    epochStruct.tm_mday = 1;
+    epochStruct.tm_mon = 0;
+    epochStruct.tm_year = 2015 - 1900;
+    epochStruct.tm_isdst = 0;
+    time_t utcEpoch = timegm(&epochStruct);
+    epoch = utcEpoch - getVariant(utcEpoch).gmtOffset;
+  }
+
+  const char* getTimezoneDirectory() {
+    const char *dir = getenv("TZDIR");
+    if (!dir) {
+      dir = DEFAULT_TZDIR;
+    }
+    return dir;
+  }
+
+  std::string getLocalTimezoneName() {
+    // use the TZ environment variable, if it is set.
+    const char *tz = getenv("TZ");
+    if (tz != nullptr) {
+      return std::string(tz);
+    }
+    // otherwise look at where /etc/localtime points and use that
+    struct stat linkStatus;
+    if (lstat(LOCAL_TIMEZONE, &linkStatus) == -1) {
+      throw TimezoneError(std::string("Can't stat local timezone link ") +
+                          LOCAL_TIMEZONE + ": " +
+                          strerror(errno));
+    }
+    std::vector<char> buffer(static_cast<size_t>(linkStatus.st_size + 1));
+    ssize_t len = readlink(LOCAL_TIMEZONE, &buffer[0], buffer.size());
+    if (len == -1 || static_cast<size_t>(len) >= buffer.size()) {
+      throw TimezoneError(std::string("Can't read local timezone link ") +
+                          LOCAL_TIMEZONE + ": " +
+                          strerror(errno));
+    }
```

Comment:
This will throw. By default, /etc/localtime is a file, not a symlink, so `readlink()` will return -1.